staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -424,6 +421,40 @@ func (c crdObjectConverter) ConvertFieldLabel(version, kind, label, value string
 	}
 }
 
+func (crdObjectConverter) Convert(in, out, context interface{}) error {
+	unstructIn, ok := in.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("input type %T in not valid for unstructured conversion", in)
+	}
+
+	unstructOut, ok := out.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("output type %T in not valid for unstructured conversion", out)
+	}
+
+	// maybe deep copy the map? It is documented in the
+	// ObjectConverter interface that this function is not
+	// guaranteeed to not mutate the input. Or maybe set the input
+	// object to nil.
+	unstructOut.Object = unstructIn.Object
+	return nil
+}
+
+func (u crdObjectConverter) ConvertToVersion(in runtime.Object, target runtime.GroupVersioner) (runtime.Object, error) {
+	gvk, ok := target.KindForGroupVersionKinds([]schema.GroupVersionKind{u.gvk})
+	if !ok {
+		// TODO: should this be a typed error?
+		return nil, fmt.Errorf("%v is unstructured and is not suitable for converting to %q", in.GetObjectKind().GroupVersionKind(), target)
+	}
+	if gvk.Version == runtime.APIVersionInternal {
+		// internal is a special case, in analogy to Scheme.ConvertToVersion.
+		gvk = schema.GroupVersionKind{}
```

Comment:
The real group version kind are lost. And we have no way to take then back.
I think this will break some code here: https://github.com/kubernetes/kubernetes/blob/676e28ff7a3fef463ac7e6801ca44855a41479ef/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go#L155

Reply:
They are not lost. The converter is per GVK, so it knows it. What is missing is a versioning serializer to set the GVK when writing to etcd.

Reply:
> What is missing is a versioning serializer to set the GVK when writing to etcd.

Get it. Thanks.